### PR TITLE
Migrated Seal and Unseal functions to tpmDirect

### PR DESCRIPTION
**Pull request by matt-tsai**
The direct implementation is back compatible: see tests.

### CODE REVIEW

File: client/keys.go [hunk 1]
```diff
@@ -450,6 +570,170 @@ func (k *Key) Unseal(in *pb.SealedBytes, opts UnsealOpts) ([]byte, error) {
...
+
+	hash := tpm.AlgID(in.GetHash())
+	sel := tpml.PCRSelection{
+		PCRSelections: []tpms.PCRSelection{
+			{
+				Hash:      hash,
+				PCRSelect: PCRs,
+			},
+		},
+	}
```
Comment (alexmwu) [inline client/keys.go:706]: this should be an unexported util function (e.g., `createTPMLPCRSelection` returning tpml.PCRSelection) in go-tpm-tools/direct since we will likely use it elsewhere and users will definitely want it at some point. Ideally, this is supported in go-tpm in a nice way, but that is out of scope for this PR.

File: client/seal_direct_test.go [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+	"github.com/google/go-tpm-tools/internal/test"
+	"github.com/google/go-tpm/direct/structures/tpm"
+	"github.com/google/go-tpm/direct/structures/tpml"
+	"github.com/google/go-tpm/direct/structures/tpms"
+	"github.com/google/go-tpm/direct/structures/tpmt"
+	tpm2direct "github.com/google/go-tpm/direct/tpm2"
+)
+
+func CreatePCRSelection(s []int) ([]byte, error) {
```
Comment (alexmwu) [inline client/seal_direct_test.go:21]: we can have this be an unexported util (e.g., `createPCRSelect`) for direct returning the []bytes that represent a PCRSelect. The util above (`createTPMLPCRSelection`) can then use this func. the two utils should go in the same util file.

File: internal/pcrs.go [hunk 3]
```diff
@@ -88,6 +91,31 @@ func SamePCRSelection(p *pb.PCRs, sel tpm2.PCRSelection) bool {
 	return true
 }
 
+// SamePCRSelectionDirect checks if the Pcrs has the same PCRSelection as the
+// provided given tpml.PCRSelection (including the hash algorithm).
```
Comment (alexmwu) [inline internal/pcrs.go:95]: nit: no need for given, just "// provided tpml.PCRSelection (including the..."

**Review by alexmwu — APPROVED** [review]
Nice work! No significant comments except pulling out duplicated code into util funcs

File: client/keys.go [hunk 4]
```diff
@@ -348,7 +360,7 @@ func sealHelper(rw io.ReadWriter, parentHandle tpmutil.Handle, auth []byte, sens
 
 	// make sure PCRs haven't being altered after sealing
 	if subtle.ConstantTimeCompare(computedDigest, decodedCreationData.PCRDigest) == 0 {
-		return nil, fmt.Errorf("PCRs have been modified after sealing")
+		return nil, fmt.Errorf("pcrs have been modified after sealing")
```
[the remark this answers is not in the document]
Comment (matt-tsai) [reply client/keys.go:363]: Resolved.

File: internal/pcrs.go (hunk 3 above)
Comment (matt-tsai) [reply internal/pcrs.go:95]: Resolved.

File: client/keys.go (hunk 1 above)
Comment (matt-tsai) [reply client/keys.go]: I added temporary unexported util functions in keys.go for tpml and tpms PCRSelections.

File: client/seal_direct_test.go (hunk 2 above)
Comment (matt-tsai) [reply client/seal_direct_test.go]: See above.

File: internal/direct_util.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+// CreateTPMLPCRSelection will create a TPMS PCR Selection given a slice of uint32 and a tpm.AlgID
+// Similar to createTPMSPCRSelection this function only supports up to 24 PCRS.
+func CreateTPMLPCRSelection(s []uint32, hash tpm.AlgID) (*tpml.PCRSelection, error) {
+	tpmsSel, err := CreateTPMSPCRSelection(s, hash)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create PCRSelection: %v", err)
+	}
+	tpmsPCR := []tpms.PCRSelection{}
```
Comment (alexmwu) [inline internal/direct_util.go:45]: nit: this can just be 
```
	sel := tpml.PCRSelection{
		PCRSelections: []tpms.PCRSelection{}{*tpmsSel},
	}
```

File: internal/direct_util.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+	"github.com/google/go-tpm/tpmutil"
+
+	"github.com/google/go-tpm/direct/structures/tpm"
+	"github.com/google/go-tpm/direct/structures/tpml"
+	"github.com/google/go-tpm/direct/structures/tpms"
+)
+
+// CreateTPMSPCRSelection will create a TPMS PCR Selection given a slice of uint32 and a tpm.AlgID.
+// This function only supports up to 24 PCRs.
+func CreateTPMSPCRSelection(s []uint32, hash tpm.AlgID) (*tpms.PCRSelection, error) {
```
Comment (alexmwu) [inline internal/direct_util.go:16]: why return a pointer to `tpms.PCRSelection`?

File: internal/direct_util.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+		Hash:      hash,
+		PCRSelect: PCRs,
+	}
+
+	return &sel, nil
+}
+
+// CreateTPMLPCRSelection will create a TPMS PCR Selection given a slice of uint32 and a tpm.AlgID
+// Similar to createTPMSPCRSelection this function only supports up to 24 PCRS.
+func CreateTPMLPCRSelection(s []uint32, hash tpm.AlgID) (*tpml.PCRSelection, error) {
```
Comment (alexmwu) [inline internal/direct_util.go:40]: why return a pointer to `tpml.PCRSelection`?

**Review by alexmwu — APPROVED** [review]

File: internal/direct_util.go (hunk 6 above)
Comment (matt-tsai) [reply internal/direct_util.go:16]: Resolved. Was following the tpmDirect format in go-tpm however it does not make sense for a higher level library to deal with unnecessary pointers.

File: internal/direct_util.go (hunk 7 above)
Comment (matt-tsai) [reply internal/direct_util.go:40]: Resolved.

File: internal/direct_util.go (hunk 5 above)
Comment (matt-tsai) [reply internal/direct_util.go:45]: Resolved.
